language/vm/src/normalized.rs

```diff
@@ -0,0 +1,288 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    access::ModuleAccess,
+    file_format::{
+        self, CompiledModule, FieldDefinition, FunctionHandle, SignatureToken, StructDefinition,
+        StructFieldInformation, TypeParameterIndex,
+    },
+};
+use move_core_types::{
+    account_address::AccountAddress,
+    identifier::Identifier,
+    language_storage::{StructTag, TypeTag},
+};
+
+/// Defines normalized representations of Move types, fields, kinds, structs, functions, and
+/// modules. These representations are useful in situations that require require comparing
+/// functions, resources, and types across modules. This arises in linking, compatibility checks
+/// (e.g., "is it safe to deploy this new module without updating its dependents and/or restarting
+/// genesis?"), defining schemas for resources stored on-chain, and (possibly in the future)
+/// allowing module updates transactions.
+
+/// A normalized version of `SignatureToken`, a type expression appearing in struct or function
+/// declarations. Unlike `SignatureToken`s, `normalized::Type`s from different modules can safely be
+/// compared.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Type {
+    Bool,
+    U8,
+    U64,
+    U128,
+    Address,
+    Signer,
+    Struct {
+        address: AccountAddress,
+        module: Identifier,
+        name: Identifier,
+        type_arguments: Vec<Type>,
+    },
+    Vector(Box<Type>),
+    TypeParameter(TypeParameterIndex),
+    Reference(Box<Type>),
+    MutableReference(Box<Type>),
+}
+
+/// Normalized version of a `FieldDefinition`. The `name` is included even though it is
+/// metadata that it is ignored by the VM. The reason: names are important to clients. We would
+/// want a change from `Account { bal: u64, seq: u64 }` to `Account { seq: u64, bal: u64 }` to be
+/// marked as incompatible. Not safe to compare without an enclosing `Struct`.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Field {
+    pub name: Identifier,
+    pub type_: Type,
+}
+
+/// Normalized version of a kind constraint on a generic type parameter. At present, this is
+/// identical to file_format::Kind, though this may change in the future.
```

Comment:
Then let's change it in the future? :) 

I think `pub use file_format::Kind` would be just as effective in keeping things separate and avoiding duplication (until we need the split)

Reply:
I don't see this code ever living somewhere it cannot refer to the `file_format`, so I don't really follow the issue

Reply:
Fixed this.